types/jsftp/index.d.ts

```diff
@@ -0,0 +1,50 @@
+// Type definitions for jsftp 2.1
+// Project: https://github.com/sergi/jsftp
+// Definitions by: Konrad Księski <https://github.com/xyleen>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+import { Socket } from 'net';
+import { EventEmitter } from 'events';
+
+export interface JsftpOpts {
+    host?: string;
+    port?: number;
+    user?: string;
+    pass?: string;
+    createSocket?: ({ port, host }: { port: number, host: string }, firstAction: () => {}) => Socket;
+    useList?: boolean;
+}
+
+export type ErrorCallback = (err: Error) => void;
+export type RawCallback = (err: Error, data: { code: number, text: string }) => void;
+export type ListCallback = (err: Error, dirContents: string) => void;
+export type GetCallback = (err: Error, socket: Socket) => void;
+export type LsCallback = (err: Error, res: [{ name: string }]) => void;
+
+export default class Ftp extends EventEmitter {
+    constructor(opts: JsftpOpts);
+
+    ls(filePath: string, callback: LsCallback): void;
+
+    list(filePath: string, callback: ListCallback): void;
+
+    get(remotePath: string, callback: GetCallback): void;
+    get(remotePath: string, localPath: string, callback: ErrorCallback): void;
+
+    put(source: string | Buffer | NodeJS.ReadableStream, remotePath: string, callback: ErrorCallback): void;
+
+    rename(from: string, to: string, callback: ErrorCallback): void;
+
+    // Ftp.raw(command, params, callback)
+    raw(command: string, callback: RawCallback): void;
```

Comment:
Per `http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html` you should use Optional Parameters in this scenario from. Please provide feedback.

Reply:
Are you reffering to `Ftp.raw()` type signatures? If so, then my reasoning is:
* the `callback` parameter, which is last, isn't optional
* `jsftp` library defines `Ftp.raw()` (https://github.com/sergi/jsftp#ftprawcommand-args-callback) as a variadic function `Ftp.raw(command, [...args], callback)`, where `args` are proxied to the FTP call with given `command`
* per SO post: https://stackoverflow.com/questions/48854207/explicit-last-function-parameter-in-typescript this might not be possible to be done in Typescript (if it is, could you please provide me with an example?)

Thanks for your feedback!

Reply:
@xyleen You're absolutely correct. After researching the `raw` command, which runs the `runCmd` function, it indeed 'automatically' picks up the last argument as the call back. It's a shame that the Rest Parameter can only be used at the end, that would make this look a lot cleaner. Thanks for your great feedback, I'll move this along.